### PR TITLE
Split encoding::types into encoding-types crate

### LINKED ISSUE #81: Extract interface into a separate crate
Some (e.g. me) might want to use `encoding`’s interfaces and build decoders and encoders for their own encodings (e.g. see #68). Pulling in the whole `encoding` with all its supported encodings and their tables is not quite ergonomic, though, and splitting interfaces/types/traits out from the main crate would keep `encoding` as simple to use as it is currently and allow people to build on it.

### PR DESCRIPTION
**Pull request by nagisa**
This allows for creation of alternative non-WHATWG encodings that use the same
interface as encodings defined in this crate without pulling in all the
tables and encodings.

This commit does not introduce any breaking changes; all the types previously defined in
encoding::types are reexported.

Fixes #81.

---

I tried to avoid breaking changes for now, but IMO `fn decode` being in `encoding::types` makes little sense; I’d move it to `encoding` at some point later.

### CODE REVIEW

**Comment by lifthrasiir** [comment]
It looks mostly fine to me, with the following caveats:

- I agree that `encoding::types::decode` is in the strange location, but we need to wait until 0.3 to get rid of it. You may move it into the crate root and reexport it from `encoding::types`.

- I guess most (and in fact *all*, but I haven't tried that) `types` tests should be actually in a new crate.

File: src/types/examples/rot13.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            Err(e) => return (0, Some(CodecError {
+                upto: 0, // Utf8Error::valid_up_to is unstable at the time of writing,
+                         // therefore we can’t quite do partial parsing yet
+                cause: "input is not valid utf-8".into()
+            }))
+        };
+        for ch in string.chars() {
+            match ch {
+                'a'...'z' => { output.write_char(rotate_byte(ch as u8) as char) },
+                'A'...'Z' => { output.write_char(rotate_byte(ch as u8) as char) },
```
Comment (lifthrasiir) [inline src/types/examples/rot13.rs]: Is this split intentional?

Comment (nagisa) [inline src/types/examples/rot13.rs]: Yes, ROT13 is defined to only work on basic latin characters (a to z and A to Z). In ASCII table uppercase and lowercase letters are not adjacent, therefore the split.

**Comment by nagisa** [comment]
> I guess most (and in fact all, but I haven't tried that) types tests should be actually in a new crate.

The tests depend on `encoding::util` which is a private module, therefore moving tests without rewriting them to not use `index_iter` is not possible. Do you still want to move them? I’ll change tests to not use `index_iter` then.

File: src/types/examples/rot13.rs (hunk 1 above)
Comment (lifthrasiir) [inline src/types/examples/rot13.rs]: I meant that `'a'...'z' | 'A'...'Z'` would probably work.

**Comment by lifthrasiir** [comment]
@nagisa Ah, that's a legitimate reason (and my suggestion can be incrementally implemented anyway). I have no reason not to merge this, so I went ahead; thank you for the PR!

**Pull request merged**